### PR TITLE
Adding path based file exclusions 

### PR DESCRIPTION
**Pull request by cpaika**
Adds the ability to exclude a path in your Vault with sensitive data from being sent to OpenAI for indexing.  Intention is to support users who want this plugin's functionality, but want to keep some files private.

Fixes https://github.com/exoascension/vault-chat/issues/20

### CODE REVIEW

File: src/VectorStore.ts [hunk 1]
```diff
@@ -274,6 +296,21 @@ export class VectorStore {
...
+		for (const key of this.embeddings.keys()) {
+			if(this.isFilePathExcluded(key)) {
+				console.log("Excluded Path:" + key)
+				keysToBeDeleted.push(key)
+			}
+		}
+		keysToBeDeleted.forEach(k => {
+			this.embeddings.delete(k)
+		})
+		await this.saveEmbeddingsToDatabaseFile() // todo debounce
```
Comment (kristenbrann) [inline src/VectorStore.ts:310]: heh we need to consider this when merging #30 which does the debouncing 😈

File: src/main.ts [hunk 2]
```diff
@@ -52,16 +53,24 @@ export default class VaultChat extends Plugin {
...
 			&& this.settings.apiKey !== DEFAULT_SETTINGS.apiKey
 			&& this.settings.apiKey.length > 30
 	}
 
+	initializeExclusion() {
+		console.log(this.settings.exclusionPath)
+		this.vectorStore.setExclusionPath(this.settings.exclusionPath)
+		this.vectorStore.deleteByPathPrefix()
+		const files = this.app.vault.getMarkdownFiles()
+		this.vectorStore.updateDatabase(files) // TODO trigger re-indexing
```
Comment (kristenbrann) [inline src/main.ts:68]: whats the todo here?

File: src/UserSettings.ts [hunk 3]
```diff
@@ -30,5 +34,19 @@ export class VaultChatSettingTab extends PluginSettingTab {
...
+
+		new Setting(containerEl)
+			.setName('Exclusion Path')
+			.setDesc('Path to exclude from OpenAI indexing. Example: `sensitive-folder/` For private files you don\'t want sent to OpenAI for training your model.')
+			.addText(text => text
+				.setValue(this.plugin.settings.exclusionPath)
+				.onChange(async (value) => this.provisionalExclusionPath = value))
+			.addButton(button => button.
+				setButtonText("Set Exclusion")
+				.onClick(async _ => {
```
Comment (kristenbrann) [inline src/UserSettings.ts:46]: what is the `_` i haven't seen this syntax

File: src/VectorStore.ts [hunk 4]
```diff
@@ -53,23 +53,40 @@ export class VectorStore {
...
 		this.tokenizer = new GPT3Tokenizer({ type: "gpt3" })
+		this.exclusionPath = exclusionPath
+	}
+
+	setExclusionPath(exclusionPath: string) {
+		this.exclusionPath = exclusionPath
+	}
+
+	private isFilePathExcluded(file: string): boolean {
+		return (!!this.exclusionPath.trim() && file.startsWith(this.exclusionPath))
```
Comment (kristenbrann) [inline src/VectorStore.ts:72]: do you know if its possible for this.exclusionPath to be null/undefined? to be safe, you may want to check that first
`this.exclusionPath && !!this.exclusionPath.trim() && file.startsWith(this.exclusionPath)`

File: src/VectorStore.ts [hunk 5]
```diff
@@ -274,6 +296,21 @@ export class VectorStore {
 		await this.saveEmbeddingsToDatabaseFile() // todo debounce
 	}
 
+	async deleteByPathPrefix() {
+		const keysToBeDeleted: string[] = []
+		for (const key of this.embeddings.keys()) {
+			if(this.isFilePathExcluded(key)) {
+				console.log("Excluded Path:" + key)
```
Comment (kristenbrann) [inline src/VectorStore.ts:303]: Debug log or remove

File: src/VectorStore.ts [hunk 6]
```diff
@@ -238,6 +255,10 @@ export class VectorStore {
 	}
 
 	async addFile(file: TFile) {
+		if(this.isFileExcluded(file)) {
+			console.log("Skipping excluded file")
```
[the remark this answers is not in the document]
Comment (cpaika) [reply src/VectorStore.ts:259]: Will just remove

File: src/main.ts (hunk 2 above)
Comment (cpaika) [reply src/main.ts:68]: Removed ty

**Review by kristenbrann — APPROVED** [review]

**Pull request merged**